### PR TITLE
Use rpmvercmp version comparison

### PR DESCRIPTION
**Pull request by jon-turney**
Addresses #3849

### CODE REVIEW

**Comment by jon-turney** [comment]
@Salamandar 

I added a comment descibing the purpose of `version_compare_condition_with_min`.  Perhaps you could just check I've got that right.

**Comment by jon-turney** [comment]
> I added a comment descibing the purpose of version_compare_condition_with_min. Perhaps you could just check I've got that right.

The behaviour seems to be wrong when the version constraint in project() doesn't establish a minimum version.  So added a small test for that and a fix.

**Comment by jon-turney** [comment]
Note that with this version comparison, the version number used for unreleased development releases e.g. '0.48.0.dev1' is greater than the released version e.g. '0.48.0' (because because it has one more element in the list, while previous elements are equal).  This is probably not desired.

**Comment by NickeZ** [comment]
Why did you chose the rpm versioning scheme and not [debian](https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-version) or [semver](https://semver.org/)? Maybe we need to support multiple sorting schemes.

Semver for example takes into account that `"0.0.1-rc1"` sorts `"<0.0.1"`.


I also don't agree with the comment below (since I'm more accustomed to semver). IMO `'0.46'` == `'0' '46' '0'` and not `'0' '46' <NULL>`. To me `'>0.46'` (=`'>0.46.0'`) would therefore mean at least `'0.46.1'` but obviously also `'0.47'`.

```
# Declaring a project(meson_version: '>0.46') and then using features in
# 0.46.0 is valid, because (knowing the meson versioning scheme) '0.46.0' is
# the lowest version which satisfies the constraint '>0.46'.
```

**Comment by jon-turney** [comment]
> Why did you chose the rpm versioning scheme and not debian or semver?

Because that apeears to be the requirement in #3849, and because it is the version comparison used by pkgconfig.

> Maybe we need to support multiple sorting schemes.

Maybe.  I think that might lead to confusion.  But this PR could be a basis for that work, if someone wanted to implement that.

> Semver for example takes into account that "0.0.1-rc1" sorts "<0.0.1".

Well, these are equal in the current 'ignore everything that isn't a . or numeric' scheme.

**Comment by jon-turney** [comment]
```
# Declaring a project(meson_version: '>0.46') and then using features in
# 0.46.0 is valid, because (knowing the meson versioning scheme) '0.46.0' is
# the lowest version which satisfies the constraint '>0.46'.
```
> In other words: I think the old behavior was right and the behavior proposed in this PR is wrong.

This comment is describing the current behaviour, and then the special steps taken to preserve it.

(But on reflection this doesn't seem consistent with how I thought the existing comparison works, so maybe I've gone wrong somewhere)

File: run_unittests.py [hunk 1]
```diff
@@ -659,6 +659,114 @@ def fake_call_pkgbin(self, args, env=None):
...
 
+    def test_version_compare(self):
+        comparefunc = mesonbuild.mesonlib.version_compare_many
+        for (a, b, result) in [
+                ('0.99.beta19', '>= 0.99.beta14', True),
+        ]:
+            self.assertEqual(comparefunc(a, b)[0], result)
+
+        for (a, b, result) in [
+                # examples from https://fedoraproject.org/wiki/Archive:Tools/RPM/VersionComparison
```
[the remark this answers is not in the document]
Comment (jon-turney) [reply run_unittests.py]: I think the ones starting at line 680 are from the RPM source code.

**Comment by Salamandar** [comment]
> Semver for example takes into account that "0.0.1-rc1" sorts "<0.0.1".

That means that requiring a dependency '>=0.0.1' will not be found, although '0.0.1-rc1' is most probably going to provide the same thing as '0.0.1'.

My guess is that '0.0.1-rc1' should be detected **just like** '0.0.1'.

File: mesonbuild/mesonlib.py [hunk 2]
```diff
@@ -440,39 +467,35 @@ def version_compare(vstr1, vstr2, strict=False):
...
-        if not version_compare(vstr1, req, strict=True):
+        if not version_compare(vstr1, req):
             not_found.append(req)
         else:
             found.append(req)
     return not_found == [], not_found, found
 
-
+# determine if the minimum version satisfying the condition |condition| exceeds
+# the minimum version for a feature |minimum|
```
Comment (Salamandar) [inline mesonbuild/mesonlib.py]: Looks ok.

**Comment by QuLogic** [comment]
A release candidate is less than the final release; it should not be detected as the same.

**Comment by jon-turney** [comment]
> (But on reflection this doesn't seem consistent with how I thought the existing comparison works, so maybe I've gone wrong somewhere)

Yeah, the example I've written in the comments doesn't make sense, as (in the current versioning scheme) a constraint in `project(meson_version:)` of `'>0.46'` is the same as `>0.46.0`, so using features added in `0.46.0` is permitted, but the `meson.build` only permits `0.46.1` or later to be used.  Writing the example with `>=` makes more sense.

**Comment by jon-turney** [comment]
My thinking is that `dependency(version:)` constraints should (by default) work just like constraints in autotools `PKG_CHECK_MODULES` checks.

It seems that `project(meson_version:)` constraints might need to work differently

**Comment by NickeZ** [comment]
[fedora packaging](https://fedoraproject.org/wiki/Packaging:Versioning) rules seems to indicate that `rpmvercmp` supports a release tag, but that tag seems to be arbitrarily chosen by the packager and doesn't seem to come from the authors of the software.

If we are to be `rpm` compatible, then it seems like we need an extra argument called "release" to distinguish correctly between `1.0` (version: `1.0`, release: `MAX_INT`(?)) and `1.0-rc0` (version: `1.0`, release: `rc0`(?)).

IMO it would be easier to standardize on semver. The biggest drawback I see is that dash (`-`) then can't be used as a "significance divider". A version called "1-0-0" would not be valid. But I wonder if this is a problem in practice.

I think the community would benefit from us nudging everyone towards a simple versioning scheme like semver. We could have more complex versioning schemes as opt-in, like `rpm:1+0-3`, `deb:1.0-3` and `[semver:]1.0.3`.

**Comment by jon-turney** [comment]
> fedora packaging rules seems to indicate that rpmvercmp supports a release tag, but that tag seems to be arbitrarily chosen by the packager and doesn't seem to come from the authors of the software.
> 
> If we are to be rpm compatible, then it seems like we need an extra argument called "release" to distinguish correctly between 1.0 (version: 1.0, release: MAX_INT(?)) and 1.0-rc0 (version: 1.0, release: rc0(?)).

No.   Stop.  

This isn't about using package identifiers.

This is about comparing the version (only) extracted from a .pc file, using the same rules as pkgconfig does.

If it makes things clearer, imagine I have written 'pkgconfig comparison'  instead of 'rpmvercmp' everywhere.

> IMO it would be easier to standardize on semver. The biggest drawback I see is that dash (-) then can't be used as a "significance divider". A version called "1-0-0" would not be valid. But I wonder if this is a problem in practice.

It absolutely is.  The problematic version strings ('0.99.beta14', '0.99.beta19') in the issue this attempts to address (#3849) aren't valid semver version strings.

I don't understand how your proposal would work in this case.

(More generally, the version strings we recieve from the dependency provider could be arbitrary ascii (at least) strings, so the default version comparison needs to give useful results in those cases, which semver doesn't even define)

> I think the community would benefit from us nudging everyone towards a simple versioning scheme like semver. We could have more complex versioning schemes as opt-in, like rpm:1+0-3, deb:1.0-3 and [semver:]1.0.3.

I think by default using a versioning scheme which is gratiously different to autotools and doesn't accept all ascii strings has more drawbacks than benefits.

File: run_unittests.py [hunk 3]
```diff
@@ -2833,6 +2959,11 @@ def test_using_recent_feature(self):
                                       ".*WARNING.*Project targetting.*but.*",
                                       meson_version='>= 0.47.0')
 
+    def test_feature_invalid_version(self):
+        os.environ['MESON_UNIT_TEST'] = '1'
+        self.assertMesonRaises("exception(mode: 'red')",
+                                "Invalid version string", meson_version=">0.44.0")
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__

[E127] continuation line over-indented for visual indent

**Comment by NickeZ** [comment]
But the example case you show must be broken, since `0.99.beta14` is > `0.99` according to your suggested sorting. When libcaca goes out of beta and they release `0.99` that version will never be selected if you have specified `>0.99.beta14`. Please correct me if I'm wrong.

**Comment by jon-turney** [comment]
> But the example case you show must be broken, since 0.99.beta14 is > 0.99 according to your suggested sorting. When libcaca goes out of beta and they release 0.99 that version will never be selected if you have specified >0.99.beta14. Please correct me if I'm wrong.

This is true.  

And if the release version is `0.99`, that will also fail to be detected as greater by `pkg-config --exists 'caca >= 0.99.beta14'` or `PKG_CHECK_MODULES([caca], [caca >= 0.99.beta14])`.  So I'd classify that as NOTOURBUG.

But I hope that the version will be at least `0.99.0`, which is greater than all of those.

**Comment by NickeZ** [comment]
Ok, I believe I've made my point so I won't argue any more.

**Comment by jon-turney** [comment]
Rebased to fix conflicts

**Comment by jon-turney** [comment]
> I've restored the version comparison previously used, for project(meson_version:) and FeatureNew checks only.
> 
> This has the property that 0.48.0.dev1 is equal to 0.48.0, and I think we have historically relied on that so that features annotated as available in 0.48.0 can be used in 0.48.0.dev1.

The motivation for this was just so habits about naming development versions don't have to change.

**Comment by jon-turney** [comment]
> Ok, I believe I've made my point so I won't argue any more.

48a7aaea8e13968e21ac6b451167c75ed85bc6b0 is as far as I got with implementing semver as an option in selectable version comparison schemes before I realized that it didn't solve my problems with the meson version number.

If you have a use for that, please feel free to finish it off..

File: mesonbuild/mesonlib.py [hunk 4]
```diff
@@ -409,14 +410,59 @@ def make_same_len(listA, listB):
...
-    if match is None:
-        msg = 'Uncomparable version string {!r}.'
-        raise MesonException(msg.format(vstr1))
-    vstr1 = match.group(0)
+        # split into numeric, alphabetic and non-alphanumeric sequences
+        sequences = re.finditer(r'(\d+|[a-zA-Z]+|[^a-zA-Z\d]+)', s)
+        # non-alphanumeric separators are discarded
+        sequences = [m for m in sequences if not re.match(r'[^a-zA-Z\d]+', m.group(1))]
+        # numeric sequences have leading zeroes discarded
+        sequences = [re.sub(r'^0+(\d)', r'\1', m.group(1), 1) for m in sequences]
```
Comment (jpakkane) [inline mesonbuild/mesonlib.py]: A more readable version of same is `txt.lstrip('0')`. This does not need to be fixed to get merged, just something to keep in mind in the future.

**Comment by jpakkane** [comment]
The reason Meson versions are of the form `0.48.0-dev1` is that it is what pip uses. However we never upload pre-release versions to pypi. Thus if we can simplify things by choosing a different numbering scheme for the pre-release dev version (such as `0.47.999`), we can do that.

File: mesonbuild/mesonlib.py (hunk 4 above)
Comment (jon-turney) [reply mesonbuild/mesonlib.py]: Good point.

But it turns out that's not quite the same.  When applied to '0', '00', '000' etc., the subsitution gives '0', whereas lstrip gives ''.

**Comment by jon-turney** [comment]
>  if we can simplify things by choosing a different numbering scheme for the pre-release dev version (such as `0.47.999`), we can do that.

Yeah, I was not very happy with the additional complexity of adding the 'meson' scheme.

Updated the PR to drop that.

**Comment by jon-turney** [comment]
Goals:

- Features annotated as available in a version should be usable in pre-release dev versions for that version

Historically, that's been fine with the 'meson' comparision scheme (as `0.48.0` == `0.48.0.dev1`).

It also works with rpmvercmp (where `0.48.0` < `0.48.0.dev1`), but would fail with some alternative pre-release version numbering considered (`0.48.0` > `0.47.9999`) or (`0.48.0` > `0.48.dev1`)

- Features used where project constraint uses a version that won't be used for an actual release

Consider `test cases/unit/41 featurenew subprojects/subprojects/bar/meson.build`:

```
project('foo subproject', meson_version: '>=0.46')
import('python')
```

where the python module first made an appearance in `0.46.0`. At the moment, this is accepted without warnings, because `0.46` == `0.46.0`  in the current version comparison (this makes implicit some knowledge about how actual releases are numbered, i.e. that `0.46` doesn't exist, but `0.46.0` does, and is the lowest possible version which satisifes the constraint)

Using rpmvercmp comparison `0.46` < `0.46.0`, so version numbers exist which satisfy the `>=0.46` constraint, but are less than the minimum version of `0.46.0` required by this feature, so we get:

```
WARNING: Project targetting '>=0.46' but tried to use feature introduced in '0.46.0': Python Module
[...]
WARNING: Project specifies a minimum meson_version '>=0.46' but uses features which were added in newer versions:
 * 0.46.0: {'Python Module'}
```

https://github.com/mesonbuild/meson/pull/4017/commits/3add9467cf53a139a8764a471e2dd5726f470e52 embeds some knowledge about how actual releases will be numbered to address this.

- Features should report the actual release version they appeared in

An alternative solution to the above would be changing the `feature_version` values in `FeatureNew` decoration from `0.46.0` to `0.46`, but then we won't report the actual release version which introduced the feature in the last line (unless we take special steps to do that...)

**Comment by jon-turney** [comment]
> The motivation for this was just so habits about naming development versions don't have to change.

I think this isn't true.  I'm not sure what I was thinking a month ago, but I believe that the simple implementation works, and meets the goals above.

**Comment by nirbheek** [comment]
Why was this merged without fixing this issue?

> WARNING: Project targetting '>= 0.47' but tried to use feature introduced in '0.47.0': User option "feature"

**Comment by nirbheek** [comment]
Ah, well, it's still happening. That's the paste from the latest master on gst-build.

**Comment by nirbheek** [comment]
This change will demonstrate the issue:

```diff
diff --git a/run_unittests.py b/run_unittests.py
index 96802cc8..d9616ea8 100755
--- a/run_unittests.py
+++ b/run_unittests.py
@@ -2973,7 +2973,7 @@ class FailureTests(BasePlatformTests):
         # Same as above, except the meson version is now appropriate
         self.assertMesonDoesNotOutput("dict = {}",
                                       ".*WARNING.*Project targetting.*but.*",
-                                      meson_version='>= 0.47.0')
+                                      meson_version='>= 0.47')

     def test_using_too_recent_feature_dependency(self):
         self.assertMesonOutputs("dependency('pcap', required: false)",
```

**Comment by jon-turney** [comment]
Doh.  The special handling in that commit doesn't work when there are spaces in the version.
Please try #4182 and see if that fixes this.

**Pull request merged**